CoreScriptsRoot/Modules/Stats/StatsAggregator.lua

```diff
@@ -0,0 +1,130 @@
+--[[
+		Filename: StatsAggregator.lua
+		Written by: dbanks
+		Description: Gather and store stats on regular heartbeat.
+--]]
+
+--[[ Services ]]--
+local CoreGuiService = game:GetService('CoreGui')
+
+--[[ Modules ]]--
+local folder = CoreGuiService:WaitForChild("RobloxGui")
+folder = folder:WaitForChild("Modules")
+folder = folder:WaitForChild("Stats")
+
+local StatsUtils = require(folder:WaitForChild( 
+    "StatsUtils"))
+
+--[[ Classes ]]--
+local StatsAggregatorClass = {}
+StatsAggregatorClass.__index = StatsAggregatorClass
+
+function StatsAggregatorClass.new(statType, numSamples, pauseBetweenSamples) 
+  local self = {}
+  setmetatable(self, StatsAggregatorClass)
+  
+  self._statType = statType
+  self._numSamples = numSamples
+  self._pauseBetweenSamples = pauseBetweenSamples
+  
+  self._statName = StatsUtils.StatNames[self._statType]
+  -- init our circular buffer.
+  self._samples = {}
+  for i = 0, numSamples-1, 1 do 
+    self._samples[i] = 0
+  end
+  self._oldestIndex = 0
+  
+  self._listeners = {}
+  
+  -- FIXME(dbanks)
+  -- Just want to be real clear this is a key, not an array index.
+  self._nextListenerId = 1001
+  
+  return self
+end
+
+function StatsAggregatorClass:AddListener(callbackFunction)
+  local id = self._nextListenerId
+  self._nextListenerId = self._nextListenerId+1
+  self._listeners[id] = callbackFunction
+  print ("New id is ", id)
+  return id
+end
+
+function StatsAggregatorClass:RemoveListener(listenerId)
+  self._listeners[listenerId] = nil
+end
+
+function StatsAggregatorClass:_notifyAllListeners()
+  for key, value in pairs(self._listeners) do
+    value()
+  end
+end
+
+function StatsAggregatorClass:StartListening()
```

Comment:
I can add a method but I am not sure when it'd be called.  

When I first go from "not showing stats" to "showing stats" it'd be nice to show a full graph of the past N measurements -> I have to aggregate even when stats aren't showing.  If it becomes a performance hit we could punt on that and just start collecting when stats are on.

But I'm only collecting once per second (or half second, something like that) so I can't imagine this is that expensive.